Overview table feature

mahoneybj opened the pull request:
## Ticket Reference
#331 

## Purpose of Changes
End user requested an overview table that was a feature in the old CO2 website

## Technical Solution
Created a new table component that can be implemented anywhere easily. Table displays rooms, readings and last updated time.

## Known Bugs

- NKB

## Personal task checklist

- [x] Meets design requirements
- [x] Has required test coverage
- [x] Addresses targeted issue and NOT a mix of multiple issues
- [x] Meets coding standards
- [x] Functions as intended
- [x] No unused code
- [x] No/minimal bugs
- [x] Decent error handling
- [x] Meets acceptance criteria

## Acceptance Criteria
- Table is clean, easy to access and read.
- Contains up to date and accurate data.
- Not overloading with information
- Checked with end users for feedback

## Current Table Design
![image](https://github.com/user-attachments/assets/e5addb7e-1740-4ad3-9590-6afe9e8d0a50)

mahoneybj commented:
Table is currently imported and rendered in Pages/BlockPage.jsx using:
```
import OverviewPanel from "../Tables/OverviewPanel";

// Render using: <OverviewPanel blockName={blockName} />
```

You can pass block Name and it will display a table rooms within that block and if block Name isn't passed it will display all sensors (Useful if we decide we want a table of all sensors).

The table can be resized easily as required depending on how the block/gages page are redesigned.

Table displays quick learning conditions using CO2 research from Liz

mahoneybj commented:
### Request for review:

## Known Bugs

- NKB

## For Testing/Review
Please tick off if the following are met (where applicable):

- [x] Meets design requirements
- [ ] Has required test coverage
- [x] Addresses targeted issue and NOT a mix of multiple issues
- [x] Meets coding standards
- [x] Functions as intended
- [x] No unused code
- [x] No/minimal bugs
- [x] Decent error handling
- [x] Meets acceptance criteria

## Acceptance Criteria
- Table is clean, easy to access and read.
- Contains up to date and accurate data.
- Not overloading with information

## Current Table Design
![image](https://github.com/user-attachments/assets/e5addb7e-1740-4ad3-9590-6afe9e8d0a50)

LizMintOS reviewed (changes requested):
## Comments

The table code is well commented, clean, and easy to follow! Your last updated code is definitely better than my implementation lol.

Anyway, there are a few changes that need to be made. I've documented many improvements in my notes, but a majority of them can be done by someone else as I'm very aware you'd like to move on to hardware again :)

## Functional Changes

**1. useGetRoomData hook is jsx.** 
> Should be js for consistency and the file doesn't return a jsx object anyway.

**2. For all your functions, move them to a reusable file and import them in your OverviewTableBody component instead.** 
> We'll definitely be using these functions elsewhere in the near future and it also adheres to best practices. 

**3. There are constant threshold values in the constants folder for indicating when a sensor is offline, etc.** 
> Can place your threshold values for all your OverviewTableBody functions there for future use and best practices. There's already a threshold value for when a sensor is offline in there, so try to use that.

**4. Questionable code in useGetRoomData hook (I don't understand it pls explain)**
> You used `lastUpdated: device.sensorData[0]?.createdAt || '',` for `if (blockName)` and `lastUpdated: device.createdAt || '',` for `else`. Why is that? Can't test the difference due to testing environment limitations, but why not use .sensorData[0] for both? 

**5. Environment should also consider temperature.** 
> Even if CO2 is 200pm, no one can work in 30 degree heat.

## Design Changes

**1. Move condition to the right of temperature.**
> The variables on the left, output on the right.

**2. Add some padding or margin between the gauges and the table.**

**3. Use colored labels for condition**
> Clearer understanding of what the different colored text means since the "condition" column tells our users what the CO2 and temp numbers mean. Easy for users to scroll through the table quickly to see which rooms are good or bad.

<img src="https://github.com/user-attachments/assets/8d63fb31-4259-4665-be51-d7a87b0f868e" width="500" />

**4. Reduce table width.**
> It's so wide that the columns are unevenly spaced and there's excess white space.

**5. Use different font weights/colors.**
> Bold should be used for the row identifier (i.e. room name) or a primary data point (like last updated, but it's already using color instead anyway)

**6. Add colors to temperature**
> Since temp is also an indicator for air quality/cognitive ability, it'd make sense to have it colored

**7. Put the table above the gauges**
> Looking at Rory's changes, it might be better to put the table above the gauges. The table should act as a summary of all the devices while the gauges (which will take up wayyyy more space than the table) provide a more visual and dedicated representation of each room.

<img src="https://github.com/user-attachments/assets/7e37f734-7c26-4313-82e4-d4630aa9995c" width="500" />

## Conclusion

@ me if you have any questions or comments!!

mahoneybj commented:
Thank you so much for that great review and suggestions, very Indepth and helpful!

### Functional changes per feedback:

1.  useGetRoomData hook is now .js instead of .jsx
2. All functions (CO2 and temperature to condition and class coloring) have been moved to ```/utils/deviceDataProcessing/deviceProcessingFunctions.js```
3. Thresholds for CO2 and temperature are moved to ```constraints.js```
4. Added comments to explain code in useGetRoomData
```
// sensorData '?' to check if the sensorData array exists to catch undefined errors
// sensorData || 0 or '' is a fallback value if the data is not available
```
While our API is setup so sensorData createdAt etc should be populated by default, good to have incase we make breaking changes etc.
5. Conditions now take into account tempature

### Design changes per feedback:

1. Temperature has been moved left before condition
2. Condition now has color badges
3. Made room number bold
4. Added color to temperature, cold is blue, green is good and hot is red

![image](https://github.com/user-attachments/assets/e00206c2-5260-4535-b5dc-6172e57b266c)


I didn't follow all design changes as the table was only put in room page for testing. I'll let Rory decide where to put it and its size when he redesigned the layout. We also talked about the option of having a toggle between table and gages etc.

Thanks for all the suggestions, I really appreciate them, I believe this feature is more well-rounded and user friendly thanks to your help!

When you get a chance I would love another review to go over my changes.

LizMintOS reviewed (changes requested):
## Comments

Looks good! The color logic is sound and moving those functions to the utils folder is a game changer.

The modular functions are generic enough for reusability and scalability. Also makes the components easier to read.

I understand that some design changes I requested were not implemented as there are some constraints at the moment, so that's all good.

## Functional Changes

**1. Rename overview hook**
> The overview hook is called `useGetRoomData` when it's getting _all room data in a given block_. I was initially confused when I first read through this function, so please rename it to be more descriptive

**2. Modular code**
> In the same hook as above, the if/else statements can be made more DRY. I'd suggest implementing something like:

```js        
const mapFn = (data) => {
          return data.map((device) => ({
            room_number: device.room_number,
            dev_eui: device.dev_eui,
            co2Level: device.sensorData[0]?.co2 || 0,
            temperature: device.sensorData[0]?.temperature || 0,
            lastUpdated: device.sensorData[0]?.createdAt || '',
            blockName: device.blockName
          }));
        };

const processedData = blockName ? mapFn(data.data.device) : mapFn(data);
```

**3. Array index logic**
> The `device.sensorData[0]` logic only takes the first sensor data in the array. I'm not sure if the `getLatestSensorData` crud route works, so try adding more seed sensor data for this block and ensure that the array index logic works as intended.


**4. Comments unnecessary in if/else statement**
> I think I caused some confusion with my previous review. When I asked why there was a difference in logic for the `useGetRoomData` code, I was actually asking for the reason behind it as, in my mind, it didn't make sense to use the code written in the `else` block. You don't need the comment in the if else statement


## Design Changes

**1. Remove color logic for CO2 and Temperature columns**
> With the new labels, there's now way too much color in the table. The Condition column clearly outlines how good each room is and why, so no need to highlight the individual data points as well.

**2. Label titles too long**
> The labels are too long to read. The table should act as a summary of the devices, so the labels should be short and sweet with their descriptions. For example, instead of "Drowsiness, headaches, poor concentration. Too cold, turn a heater", make it shorter like "Cold and stuffy" OR display two labels in one column to describe temp and co2. E.g. "Cold" "Poor air quality". Try out these two routes!

**3. Too much space**
> Can reduce the space between the conditions column and last updated, if that's possible.

## Conclusion

I understand that you want to move on from software onto hardware ASAP! Once these changes are made, I'm happy to approve. There are definitely some more changes to be made, but I can work on that as part of my current table issue or in a future iteration.

Let me know if you have any questions or feedback. I'm happy to work with you on this.
@mahoneybj

mahoneybj commented:
### Request for rereview

Hi Liz,
Thanks for the review and great suggestions and catches!

Changes:

- Renamed hook to ```useGetRoomSensorsByBlock```
- Better coding practices in ```useGetRoomSensorsByBlock```
- Removed unnecessary comments and code
- Removed colour from CO2 and Temp coloums
- Shortened the condition labels
- Reduced table width

Let me know if you have any questions.

LizMintOS reviewed (approved):
Looks good, Ben! 

All the requested changes have been made and I'm happy with how they look at this time.

As discussed, the Overview table will be unused until Rory's feature is implemented and the layout is sorted. Great work in the mean time!
